src/core/stdc/clang_block.d

```diff
@@ -0,0 +1,198 @@
+ *
+ * Examples:
+ *
+ * Basic example of calling a function taking a block with no arguments that
+ * returns void.
+ *
+ * $(CCODE void foo(void (^block)(void));)
+ *
+ * Will look like:
+ *
+ * ---
+ * import core.stdc.clang_block;
+ *
+ * // The `Block` type is used to represent the Clang block in D
+ * extern(C) void foo(Block!()* block);
+ *
+ * void main()
+ * {
+ *      // The `block` function is used to initialize an instance of `Block`.
+ *      // A delegate will be passed to the `block` function which will be the
+ *      // body of the block.
+ *      auto b = block({ writeln("foo"); });
+ *      foo(&b);
+ * }
+ * ---
+ *
+ * Example of calling a function taking a block with arguments that returns
+ * void.
+ *
+ * $(CCODE void foo(void (^block)(int));)
+ *
+ * Will look like:
+ *
+ * ---
+ * import core.stdc.clang_block;
+ *
+ * // The type parameters to the instantiation of `Block` is first the return
+ * // type, `void`, and then the parameter types, `int`.
+ * extern(C) void foo(Block!(void, int)* block);
+ *
+ * void main()
+ * {
+ *      // The delegate to the `block` function cannot use inferred type
+ *      // parameters.
+ *      auto b = block((int a){ writeln(a); });
+ *      foo(&b);
+ * }
+ * ---
+ *
+ * Example of calling a function taking a block with no arguments that returns
+ * an int.
+ *
+ * $(CCODE void foo(int (^block)(void));)
+ *
+ * Will look like:
+ *
+ * ---
+ * import core.stdc.clang_block;
+ *
+ * extern(C) void foo(Block!(int)* block);
+ *
+ * void main()
+ * {
+ *      // The return type can be inferred
+ *      auto b = block({ return 3; });
+ *      foo(&b);
+ * }
+ * ---
+ *
+ * Copyright: Copyright Jacob Carlborg 2016.
+ * License:   $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:   Jacob Carlborg
+ * Source:    $(DRUNTIMESRC core/_clang_block.d)
+ */
+module core.stdc.clang_block;
+
+import core.stdc.config;
+
+version(CoreDdoc)
+{
+    /**
+     * This struct is the D representation of a Clang Block.
+     *
+     * Params:
+     *  R = the return type of the block
+     *  Params = the parameter types of the block
+     */
+    struct Block(R = void, Params...)
+    {
+    }
+
+    /**
+     * Creates a new block that can be passed to a C function expecting a
+     * Clang Block.
+     *
+     * Params:
+     *  R = the return type of the block
+     *  Params = the parameter types of the block
+     *  dg = the body of the block
+     *
+     * Returns: the newly created block
+     */
+    Block!(R, Params) block(R, Params...)(R delegate(Params) dg)
+    {
+        return Block.init;
+    }
+}
+
+else
+    version(OSX):
+
+struct Block(R = void, Params...)
+{
+private:
+    void* isa;
+    int flags;
+    int reserved;
+    extern(C) R function(Block*, Params) invoke;
+    Descriptor* descriptor;
+
+    // Imported variables go here
+    R delegate(Params) dg;
+}
+
+Block!(R, Params) block(R, Params...)(R delegate(Params) dg)
```

Comment:
Both would work, though the `union` will complicate things. It's up to you if you want to support it, just though I would raise the point.